services/autoscaling/src/simcore_service_autoscaling/utils_docker.py

```diff
@@ -2,163 +2,219 @@
-    """
-
     async with aiodocker.Docker() as docker:
-        nodes = await docker.nodes.list(filters={"label": "sidecar"})
-        nodes_sidecar_data = []
-        for node in nodes:
-            nodes_sidecar_data.append(
-                {
-                    "ID": node["ID"],
-                    "RAM": bytesto(
-                        node["Description"]["Resources"]["MemoryBytes"],
-                        "g",
-                        bsize=1024,
-                    ),
-                    "CPU": int(node["Description"]["Resources"]["NanoCPUs"])
-                    / 1000000000,
+        tasks = parse_obj_as(
+            list[Task],
+            await docker.tasks.list(
+                filters={
+                    "desired-state": "running",
+                    "label": service_labels,
                 }
-            )
-
-        total_nodes_cpus = 0
-        total_nodes_ram = 0
-        nodes_ids = []
-        for node in nodes_sidecar_data:
-            total_nodes_cpus += node["CPU"]
-            total_nodes_ram += node["RAM"]
-            nodes_ids.append(node["ID"])
-
-        return ClusterResources(
-            total_cpus=total_nodes_cpus,
-            total_ram=total_nodes_ram,
-            nodes_ids=nodes_ids,
+            ),
         )
 
+    def _is_task_waiting_for_resources(task: Task) -> bool:
+        if (
+            not task.Status
+            or not task.Status.State
+            or not task.Status.Message
+            or not task.Status.Err
+        ):
+            return False
+        return (
+            task.Status.State == TaskState.pending
+            and task.Status.Message == "pending task scheduling"
+            and "insufficient resources on" in task.Status.Err
+        )
 
-class TasksResources(BaseModel):
-    total_cpus_running_tasks: PositiveInt
-    total_ram_running_tasks: PositiveInt
-    total_cpus_pending_tasks: PositiveInt
-    total_ram_pending_tasks: PositiveInt
-    count_tasks_pending: PositiveInt
-
+    pending_tasks = [task for task in tasks if _is_task_waiting_for_resources(task)]
 
-class ReservedResources(TypedDict):
-    RAM: float
-    CPU: int
-    # NOTE: future add VRAM, AIRAM
-    # SEE https://github.com/ITISFoundation/osparc-simcore/pull/3364#discussion_r987821694
+    return pending_tasks
 
 
-def _get_reserved_resources(reservations: dict[str, Any]) -> ReservedResources:
-    ram: float = 0.0
-    cpu: int = 0
-    if "MemoryBytes" in reservations:
-        ram = bytesto(
-            reservations["MemoryBytes"],
-            "g",
-            bsize=1024,
+async def compute_cluster_total_resources(nodes: list[Node]) -> Resources:
+    """
+    Returns the nodes total resources.
+    """
+    cluster_resources_counter = collections.Counter({"ram": 0, "cpus": 0})
+    for node in nodes:
+        assert node.Description  # nosec
+        assert node.Description.Resources  # nosec
+        assert node.Description.Resources.NanoCPUs  # nosec
```

Comment:
Are these always present? Not sure I like these asserts.

Reply:
These are coming from the fact that the generated pydantic model is overusing Optional keywords...